File: tests/shm_test_client.c
/*
 * Minimal SHM-based Wayland Test Client
 * 
 * This program creates a simple window and renders content using shared memory.
 * Used to test the Axiom compositor's rendering pipeline.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <errno.h>
#include <wayland-client.h>
#include <wayland-client-protocol.h>

/* XDG Shell protocol - you'll need to generate these from the XML */
#include "xdg-shell-client-protocol.h"

#define WIDTH 800
#define HEIGHT 600

struct client_state {
    struct wl_display *display;
    struct wl_registry *registry;
    struct wl_compositor *compositor;
    struct wl_shm *shm;
    struct xdg_wm_base *xdg_wm_base;
    struct wl_surface *surface;
    struct xdg_surface *xdg_surface;
    struct xdg_toplevel *xdg_toplevel;
    struct wl_buffer *buffer;
    struct wl_shm_pool *pool;
    void *shm_data;
    int running;
    int configured;
};

/* Create anonymous file for shared memory */
static int create_shm_file(size_t size) {
    char name[256];
    int fd;
    
    snprintf(name, sizeof(name), "/axiom-shm-test-%d", getpid());
    
    fd = shm_open(name, O_RDWR | O_CREAT | O_EXCL, 0600);
    if (fd < 0) {
        fprintf(stderr, "shm_open failed: %s\n", strerror(errno));
        return -1;
    }
    
    shm_unlink(name);
    
    if (ftruncate(fd, size) < 0) {
        fprintf(stderr, "ftruncate failed: %s\n", strerror(errno));
        close(fd);
        return -1;
    }
    
    return fd;
}

/* Draw a test pattern to the buffer */
static void draw_test_pattern(uint32_t *pixels, int width, int height) {
    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            uint32_t color;
            
            /* Create a gradient + checkerboard pattern */
            int checker = ((x / 32) + (y / 32)) % 2;
            
            if (checker) {
                /* Red gradient */
                uint8_t r = (x * 255) / width;
                uint8_t g = 50;
                uint8_t b = 50;
                color = 0xFF000000 | (r << 16) | (g << 8) | b;
            } else {
                /* Blue gradient */
                uint8_t r = 50;
                uint8_t g = (y * 255) / height;
                uint8_t b = 200;
                color = 0xFF000000 | (r << 16) | (g << 8) | b;
            }
            
            pixels[y * width + x] = color;
        }
    }
    
    printf("✅ Drew test pattern: %dx%d pixels\n", width, height);
}

/* Create shared memory buffer */
static struct wl_buffer *create_buffer(struct client_state *state, int width, int height) {
    int stride = width * 4; /* 4 bytes per pixel (ARGB8888) */
    int size = stride * height;
    
    /* Create shared memory file */
    int fd = create_shm_file(size);
    if (fd < 0) {
        return NULL;
    }
    
    /* Map the memory */
    void *data = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    if (data == MAP_FAILED) {
        fprintf(stderr, "mmap failed: %s\n", strerror(errno));
        close(fd);
        return NULL;
    }
    
    state->shm_data = data;
    
    /* Draw the test pattern */
    draw_test_pattern((uint32_t *)data, width, height);
    
    /* Create wl_shm_pool */
    state->pool = wl_shm_create_pool(state->shm, fd, size);
    close(fd);
    
    if (!state->pool) {
        fprintf(stderr, "Failed to create wl_shm_pool\n");
        munmap(data, size);
        return NULL;
    }
    
    /* Create wl_buffer from pool */
    struct wl_buffer *buffer = wl_shm_pool_create_buffer(
        state->pool, 0, width, height, stride, WL_SHM_FORMAT_ARGB8888
    );
    
    if (!buffer) {
        fprintf(stderr, "Failed to create wl_buffer\n");
        wl_shm_pool_destroy(state->pool);
        munmap(data, size);
        return NULL;
    }
    
    printf("✅ Created SHM buffer: %dx%d, stride=%d, size=%d bytes\n", 
           width, height, stride, size);
    
    return buffer;
}

/* XDG Surface configure handler */
static void xdg_surface_configure(void *data, struct xdg_surface *xdg_surface, uint32_t serial) {
    struct client_state *state = data;
    xdg_surface_ack_configure(xdg_surface, serial);
    
    state->configured = 1;
    printf("✅ XDG surface configured (serial=%u)\n", serial);
    
    /* On first configure, attach buffer and commit */
    if (state->buffer && state->surface) {
        wl_surface_attach(state->surface, state->buffer, 0, 0);
        wl_surface_damage(state->surface, 0, 0, WIDTH, HEIGHT);
        wl_surface_commit(state->surface);
        printf("✅ Attached buffer and committed surface\n");
    }
}

static const struct xdg_surface_listener xdg_surface_listener = {
    .configure = xdg_surface_configure,
};

/* XDG Toplevel configure handler */
static void xdg_toplevel_configure(void *data, struct xdg_toplevel *xdg_toplevel,
                                   int32_t width, int32_t height, struct wl_array *states) {
    printf("ℹ️  Toplevel configure: %dx%d\n", width, height);
}

static void xdg_toplevel_close(void *data, struct xdg_toplevel *xdg_toplevel) {
    struct client_state *state = data;
    state->running = 0;
    printf("🚪 Window close requested\n");
}

static const struct xdg_toplevel_listener xdg_toplevel_listener = {
    .configure = xdg_toplevel_configure,
    .close = xdg_toplevel_close,
};

/* XDG WM Base ping handler */
static void xdg_wm_base_ping(void *data, struct xdg_wm_base *xdg_wm_base, uint32_t serial) {
    xdg_wm_base_pong(xdg_wm_base, serial);
}

static const struct xdg_wm_base_listener xdg_wm_base_listener = {
    .ping = xdg_wm_base_ping,
};

/* Registry global handler */
static void registry_global(void *data, struct wl_registry *registry,
                           uint32_t name, const char *interface, uint32_t version) {
    struct client_state *state = data;
    
    printf("📋 Registry: %s (id=%u, version=%u)\n", interface, name, version);
    
    if (strcmp(interface, "wl_compositor") == 0) {
        state->compositor = wl_registry_bind(registry, name, &wl_compositor_interface, 4);
        printf("✅ Bound wl_compositor\n");
    } else if (strcmp(interface, "wl_shm") == 0) {
        state->shm = wl_registry_bind(registry, name, &wl_shm_interface, 1);
        printf("✅ Bound wl_shm\n");
    } else if (strcmp(interface, "xdg_wm_base") == 0) {
        state->xdg_wm_base = wl_registry_bind(registry, name, &xdg_wm_base_interface, 1);
        xdg_wm_base_add_listener(state->xdg_wm_base, &xdg_wm_base_listener, state);
        printf("✅ Bound xdg_wm_base\n");
    }
}

static void registry_global_remove(void *data, struct wl_registry *registry, uint32_t name) {
    /* Not used */
}

static const struct wl_registry_listener registry_listener = {
    .global = registry_global,
    .global_remove = registry_global_remove,
};

int main(int argc, char *argv[]) {
    struct client_state state = {0};
    state.running = 1;
    
    printf("🚀 Starting Axiom SHM Test Client\n");
    printf("================================\n\n");
    
    /* Connect to Wayland display */
    state.display = wl_display_connect(NULL);
    if (!state.display) {
        fprintf(stderr, "❌ Failed to connect to Wayland display\n");
        return 1;
    }
    printf("✅ Connected to Wayland display\n");
    
    /* Get registry and bind globals */
    state.registry = wl_display_get_registry(state.display);
    wl_registry_add_listener(state.registry, &registry_listener, &state);
    wl_display_roundtrip(state.display);
    
    /* Check that we got all required interfaces */
    if (!state.compositor || !state.shm || !state.xdg_wm_base) {
        fprintf(stderr, "❌ Missing required Wayland interfaces\n");
        fprintf(stderr, "   compositor: %p, shm: %p, xdg_wm_base: %p\n",
                state.compositor, state.shm, state.xdg_wm_base);
        return 1;
    }
    
    printf("\n📐 Creating window (%dx%d)\n", WIDTH, HEIGHT);
    
    /* Create surface */
    state.surface = wl_compositor_create_surface(state.compositor);
    if (!state.surface) {
        fprintf(stderr, "❌ Failed to create wl_surface\n");
        return 1;
    }
    printf("✅ Created wl_surface\n");
    
    /* Create XDG surface */
    state.xdg_surface = xdg_wm_base_get_xdg_surface(state.xdg_wm_base, state.surface);
    if (!state.xdg_surface) {
        fprintf(stderr, "❌ Failed to create xdg_surface\n");
        return 1;
    }
    xdg_surface_add_listener(state.xdg_surface, &xdg_surface_listener, &state);
    printf("✅ Created xdg_surface\n");
    
    /* Create XDG toplevel */
    state.xdg_toplevel = xdg_surface_get_toplevel(state.xdg_surface);
    if (!state.xdg_toplevel) {
        fprintf(stderr, "❌ Failed to create xdg_toplevel\n");
        return 1;
    }
    xdg_toplevel_add_listener(state.xdg_toplevel, &xdg_toplevel_listener, &state);
    xdg_toplevel_set_title(state.xdg_toplevel, "Axiom SHM Test");
    printf("✅ Created xdg_toplevel\n");
    
    /* Create shared memory buffer */
    printf("\n🎨 Creating SHM buffer\n");
    state.buffer = create_buffer(&state, WIDTH, HEIGHT);
    if (!state.buffer) {
        fprintf(stderr, "❌ Failed to create buffer\n");
        return 1;
    }
    
    /* Initial commit to map the window */
    wl_surface_commit(state.surface);
    printf("✅ Committed initial surface\n");
    
    /* Wait for configure */
    printf("\n⏳ Waiting for configure event...\n");
    while (!state.configured && state.running) {
        if (wl_display_dispatch(state.display) == -1) {
            fprintf(stderr, "❌ Display dispatch failed\n");
            break;
        }
    }
    
    if (state.configured) {
        printf("\n✨ Window is now visible and should display test pattern!\n");
        printf("   - Red/blue checkerboard with gradients\n");
        printf("   - Press Ctrl+C to exit\n\n");
    }
    
    /* Main event loop */
    printf("🔄 Entering main loop...\n");
    while (state.running) {
        if (wl_display_dispatch(state.display) == -1) {
            fprintf(stderr, "❌ Display dispatch failed\n");
            break;
        }
    }
    
    /* Cleanup */
    printf("\n🧹 Cleaning up...\n");
    
    if (state.buffer) wl_buffer_destroy(state.buffer);
    if (state.pool) wl_shm_pool_destroy(state.pool);
    if (state.shm_data) munmap(state.shm_data, WIDTH * HEIGHT * 4);
    if (state.xdg_toplevel) xdg_toplevel_destroy(state.xdg_toplevel);
    if (state.xdg_surface) xdg_surface_destroy(state.xdg_surface);
    if (state.surface) wl_surface_destroy(state.surface);
    if (state.xdg_wm_base) xdg_wm_base_destroy(state.xdg_wm_base);
    if (state.compositor) wl_compositor_destroy(state.compositor);
    if (state.shm) wl_shm_destroy(state.shm);
    if (state.registry) wl_registry_destroy(state.registry);
    if (state.display) wl_display_disconnect(state.display);
    
    printf("✅ Shutdown complete\n");
    return 0;
}

File: tests/shm_test_client.py
#!/usr/bin/env python3
"""
Axiom SHM Test Client (Python version)

This is a Python-based Wayland client that uses shared memory (SHM) buffers
to test the Axiom compositor's rendering pipeline.

Requirements:
    pip install pywayland

Usage:
    python3 shm_test_client.py
"""

import os
import sys
import mmap
import tempfile
import struct
import signal
from pywayland.client import Display
from pywayland.protocol.wayland import (
    WlCompositor,
    WlShm,
    WlShmPool,
    WlBuffer,
    WlSurface,
)
from pywayland.protocol.xdg_shell import XdgWmBase, XdgSurface, XdgToplevel

# Window dimensions
WIDTH = 800
HEIGHT = 600
BYTES_PER_PIXEL = 4  # ARGB8888


class ShmTestClient:
    """Simple SHM-based Wayland test client"""

    def __init__(self):
        self.display = None
        self.compositor = None
        self.shm = None
        self.xdg_wm_base = None
        self.surface = None
        self.xdg_surface = None
        self.xdg_toplevel = None
        self.buffer = None
        self.shm_pool = None
        self.shm_data = None
        self.shm_fd = None
        self.configured = False
        self.running = True

    def create_shm_buffer(self, width, height):
        """Create a shared memory buffer and draw test pattern"""
        stride = width * BYTES_PER_PIXEL
        size = stride * height

        print(
            f"📐 Creating SHM buffer: {width}x{height}, stride={stride}, size={size} bytes"
        )

        # Create anonymous file
        fd = os.memfd_create("axiom-shm-test", os.MFD_CLOEXEC)
        os.ftruncate(fd, size)
        self.shm_fd = fd

        # Map the memory
        self.shm_data = mmap.mmap(
            fd, size, mmap.MAP_SHARED, mmap.PROT_READ | mmap.PROT_WRITE
        )

        # Draw test pattern
        print("🎨 Drawing test pattern...")
        self.draw_test_pattern(width, height, stride)

        # Create wl_shm_pool
        self.shm_pool = self.shm.create_pool(fd, size)

        # Create wl_buffer
        self.buffer = self.shm_pool.create_buffer(
            0,  # offset
            width,
            height,
            stride,
            WlShm.format.argb8888.value,
        )

        print(f"✅ SHM buffer created successfully")
        return self.buffer

    def draw_test_pattern(self, width, height, stride):
        """Draw a colorful test pattern to the buffer"""
        for y in range(height):
            for x in range(width):
                # Create a gradient + checkerboard pattern
                checker = ((x // 32) + (y // 32)) % 2

                if checker:
                    # Red gradient
                    r = int((x * 255) / width)
                    g = 50
                    b = 50
                else:
                    # Blue gradient
                    r = 50
                    g = int((y * 255) / height)
                    b = 200

                # ARGB8888 format (little endian)
                alpha = 0xFF
                color = (alpha << 24) | (r << 16) | (g << 8) | b

                # Write to buffer
                offset = y * stride + x * BYTES_PER_PIXEL
                struct.pack_into("<I", self.shm_data, offset, color)

        print(f"✅ Drew test pattern: {width}x{height} pixels")

    def handle_registry_global(self, registry, id, interface, version):
        """Handle registry global events"""
        print(f"📋 Registry: {interface} (id={id}, version={version})")

        if interface == "wl_compositor":
            self.compositor = registry.bind(id, WlCompositor, version)
            print("✅ Bound wl_compositor")
        elif interface == "wl_shm":
            self.shm = registry.bind(id, WlShm, version)
            print("✅ Bound wl_shm")
        elif interface == "xdg_wm_base":
            self.xdg_wm_base = registry.bind(id, XdgWmBase, version)
            self.xdg_wm_base.dispatcher["ping"] = self.handle_xdg_wm_base_ping
            print("✅ Bound xdg_wm_base")

    def handle_registry_global_remove(self, registry, id):
        """Handle registry global remove events"""
        pass

    def handle_xdg_wm_base_ping(self, xdg_wm_base, serial):
        """Handle XDG WM Base ping"""
        xdg_wm_base.pong(serial)

    def handle_xdg_surface_configure(self, xdg_surface, serial):
        """Handle XDG surface configure"""
        xdg_surface.ack_configure(serial)
        self.configured = True
        print(f"✅ XDG surface configured (serial={serial})")

        # On first configure, attach buffer and commit
        if self.buffer and self.surface and not self.configured:
            self.surface.attach(self.buffer, 0, 0)
            self.surface.damage(0, 0, WIDTH, HEIGHT)
            self.surface.commit()
            print("✅ Attached buffer and committed surface")

    def handle_xdg_toplevel_configure(self, xdg_toplevel, width, height, states):
        """Handle XDG toplevel configure"""
        if width > 0 and height > 0:
            print(f"ℹ️  Toplevel configure: {width}x{height}")

    def handle_xdg_toplevel_close(self, xdg_toplevel):
        """Handle XDG toplevel close"""
        print("🚪 Window close requested")
        self.running = False

    def setup_signal_handlers(self):
        """Setup signal handlers for clean shutdown"""

        def signal_handler(signum, frame):
            print("\n🛑 Received signal, shutting down...")
            self.running = False

        signal.signal(signal.SIGINT, signal_handler)
        signal.signal(signal.SIGTERM, signal_handler)

    def run(self):
        """Main client loop"""
        print("🚀 Starting Axiom SHM Test Client (Python)")
        print("=" * 60)
        print()

        self.setup_signal_handlers()

        # Connect to Wayland display
        try:
            self.display = Display()
            self.display.connect()
            print("✅ Connected to Wayland display")
        except Exception as e:
            print(f"❌ Failed to connect to Wayland display: {e}")
            return 1

        # Get registry and bind globals
        registry = self.display.get_registry()
        registry.dispatcher["global"] = self.handle_registry_global
        registry.dispatcher["global_remove"] = self.handle_registry_global_remove

        # Process registry
        self.display.dispatch(block=True)
        self.display.roundtrip()

        # Check that we got required interfaces
        if not self.compositor or not self.shm or not self.xdg_wm_base:
            print("❌ Missing required Wayland interfaces")
            print(f"   compositor: {self.compositor}")
            print(f"   shm: {self.shm}")
            print(f"   xdg_wm_base: {self.xdg_wm_base}")
            return 1

        print()
        print(f"📐 Creating window ({WIDTH}x{HEIGHT})")

        # Create surface
        self.surface = self.compositor.create_surface()
        print("✅ Created wl_surface")

        # Create XDG surface
        self.xdg_surface = self.xdg_wm_base.get_xdg_surface(self.surface)
        self.xdg_surface.dispatcher["configure"] = self.handle_xdg_surface_configure
        print("✅ Created xdg_surface")

        # Create XDG toplevel
        self.xdg_toplevel = self.xdg_surface.get_toplevel()
        self.xdg_toplevel.dispatcher["configure"] = self.handle_xdg_toplevel_configure
        self.xdg_toplevel.dispatcher["close"] = self.handle_xdg_toplevel_close
        self.xdg_toplevel.set_title("Axiom SHM Test (Python)")
        print("✅ Created xdg_toplevel")

        # Create SHM buffer
        print()
        print("🎨 Creating SHM buffer")
        try:
            self.create_shm_buffer(WIDTH, HEIGHT)
        except Exception as e:
            print(f"❌ Failed to create SHM buffer: {e}")
            import traceback

            traceback.print_exc()
            return 1

        # Initial commit to map the window
        self.surface.commit()
        print("✅ Committed initial surface")

        # Wait for configure
        print()
        print("⏳ Waiting for configure event...")
        while not self.configured and self.running:
            self.display.dispatch(block=True)

        if self.configured:
            # Attach buffer after configure
            self.surface.attach(self.buffer, 0, 0)
            self.surface.damage(0, 0, WIDTH, HEIGHT)
            self.surface.commit()

            print()
            print("✨ Window is now visible and should display test pattern!")
            print("   - Red/blue checkerboard with gradients")
            print("   - Press Ctrl+C to exit")
            print()

        # Main event loop
        print("🔄 Entering main loop...")
        while self.running:
            try:
                self.display.dispatch(block=True)
            except Exception as e:
                print(f"❌ Display dispatch error: {e}")
                break

        # Cleanup
        print()
        print("🧹 Cleaning up...")
        self.cleanup()

        print("✅ Shutdown complete")
        return 0

    def cleanup(self):
        """Clean up resources"""
        if self.buffer:
            self.buffer.destroy()
        if self.shm_pool:
            self.shm_pool.destroy()
        if self.shm_data:
            self.shm_data.close()
        if self.shm_fd:
            os.close(self.shm_fd)
        if self.xdg_toplevel:
            self.xdg_toplevel.destroy()
        if self.xdg_surface:
            self.xdg_surface.destroy()
        if self.surface:
            self.surface.destroy()
        if self.xdg_wm_base:
            self.xdg_wm_base.destroy()
        if self.compositor:
            self.compositor.destroy()
        if self.shm:
            self.shm.destroy()
        if self.display:
            self.display.disconnect()


def main():
    """Main entry point"""
    # Check for pywayland
    try:
        import pywayland

        print(f"ℹ️  Using pywayland version: {pywayland.__version__}")
    except ImportError:
        print("❌ pywayland not found")
        print("   Install with: pip install pywayland")
        return 1

    # Check for memfd_create (Python 3.8+)
    if not hasattr(os, "memfd_create"):
        print("❌ os.memfd_create not available (Python 3.8+ required)")
        return 1

    # Run the client
    client = ShmTestClient()
    try:
        return client.run()
    except KeyboardInterrupt:
        print("\n🛑 Interrupted by user")
        client.cleanup()
        return 0
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        import traceback

        traceback.print_exc()
        client.cleanup()
        return 1


if __name__ == "__main__":
    sys.exit(main())
